File: kythe/cxx/extractor/testdata/root_directory_test.go
package root_directory_test

import (
	"fmt"
	"log"
	"testing"

	"os"
	"os/exec"
	"path"

	"github.com/bazelbuild/rules_go/go/tools/bazel"
	"github.com/golang/protobuf/proto"
	"github.com/golang/protobuf/ptypes"
	"github.com/golang/protobuf/ptypes/any"
	"github.com/google/go-cmp/cmp"

	"kythe.io/kythe/go/platform/kzip"

	apb "kythe.io/kythe/proto/analysis_go_proto"
	fcpb "kythe.io/kythe/proto/filecontext_go_proto"
	spb "kythe.io/kythe/proto/storage_go_proto"
)

var (
	testName         = "root_directory_test"
	extractorRunpath = "kythe/cxx/extractor/cxx_extractor"
	fileRunpath      = "kythe/cxx/extractor/testdata/altroot/altpath/file.cc"
	outputPath       = path.Join(os.Getenv("TEST_TMPDIR"), testName+".kzip")
	expectedUnit     = apb.CompilationUnit{
		VName: &spb.VName{
			Language: "c++",
		},
		RequiredInput: []*apb.CompilationUnit_FileInput{
			{
				VName: &spb.VName{Path: "altpath/file.cc"},
				Info: &apb.FileInfo{
					Path:   "file.cc",
					Digest: "a24091884bc15b53e380fe5b874d1bb52d89269fdf2592808ac70ba189204730",
				},
				Details: []*any.Any{
					mustMarshal(&fcpb.ContextDependentVersion{
						Row: []*fcpb.ContextDependentVersion_Row{
							{SourceContext: "hash0"},
						},
					}),
				},
			},
		},
		Argument: []string{
			"/dummy/bin/g++",
			"-target",
			"dummy-target",
			"-DKYTHE_IS_RUNNING=1",
			"-resource-dir",
			"/kythe_builtins",
			"--driver-mode=g++",
			"file.cc",
			"-fsyntax-only",
		},
		SourceFile:   []string{"file.cc"},
		EntryContext: "hash0",
	}
)

func TestRootDirectory(t *testing.T) {
	defer os.RemoveAll(outputPath)
	extractor, err := bazel.Runfile(extractorRunpath)
	if err != nil {
		t.Fatal(err)
	}
	filepath, err := bazel.Runfile(fileRunpath)
	if err != nil {
		t.Fatal(err)
	}
	cmd := exec.Command(extractor, "--with_executable", "/dummy/bin/g++", path.Base(filepath))
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	cmd.Dir = path.Dir(filepath)
	cmd.Env = append(os.Environ(),
		"KYTHE_EXCLUDE_EMPTY_DIRS=1",
		"KYTHE_EXCLUDE_AUTOCONFIGURATION_FILES=1",
		"KYTHE_ROOT_DIRECTORY="+path.Dir(path.Dir(filepath)),
		"KYTHE_OUTPUT_FILE="+outputPath)
	if err := cmd.Run(); err != nil {
		t.Fatal(err)
	}
	units, err := readAll(outputPath)
	if err != nil {
		t.Fatal(err)
	}
	if len(units) != 1 {
		t.Fatalf("unexpected number of compilation units: got %d, want %d", len(units), 1)
	}

	// Remove or normalize fields that are likely to differ between runs.
	units[0].Proto.VName.Signature = ""
	units[0].Proto.WorkingDirectory = path.Clean(units[0].Proto.WorkingDirectory)
	units[0].Proto.Argument[2] = "dummy-target"
	units[0].Proto.Details = nil

	if err := canonicalizeContextHashes(units[0].Proto); err != nil {
		t.Fatal(err)
	}

	expectedUnit.WorkingDirectory = path.Clean(cmd.Dir)
	if diff := cmp.Diff(expectedUnit, *units[0].Proto); diff != "" {
		t.Fatalf("unexpected VName differences: (- expected; + found)\n%s", diff)
	}
}

func mustMarshal(msg proto.Message) *any.Any {
	any, err := ptypes.MarshalAny(msg)
	if err != nil {
		log.Fatal(err)
	}
	return any
}

func readAll(path string) ([]*kzip.Unit, error) {
	file, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	var result []*kzip.Unit
	return result, kzip.Scan(file, func(_ *kzip.Reader, unit *kzip.Unit) error {
		result = append(result, unit)
		return nil
	})
}

func canonicalizeHash(hashes map[string]string, hash string) string {
	if hash == "" {
		return hash
	}
	size := len(hashes)
	found, ok := hashes[hash]
	if ok {
		return found
	}
	found = "hash" + fmt.Sprint(size)
	hashes[hash] = found
	return found
}

func canonicalizeContextHashes(unit *apb.CompilationUnit) error {
	hashes := make(map[string]string)
	unit.EntryContext = canonicalizeHash(hashes, unit.EntryContext)
	for _, input := range unit.RequiredInput {
		for i, any := range input.Details {
			var version fcpb.ContextDependentVersion
			if ptypes.Is(any, &version) {
				if err := ptypes.UnmarshalAny(any, &version); err != nil {
					return err
				}
			}
			for _, row := range version.Row {
				row.SourceContext = canonicalizeHash(hashes, row.SourceContext)
				for _, column := range row.Column {
					column.LinkedContext = canonicalizeHash(hashes, column.LinkedContext)
				}
			}
			any, err := ptypes.MarshalAny(&version)
			if err != nil {
				return err
			}
			input.Details[i] = any
		}
	}
	return nil
}

File: kythe/cxx/extractor/testdata/root_directory_test.cc
/*
 * Copyright 2019 The Kythe Authors. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <optional>
#include <string>

#include "absl/strings/string_view.h"
#include "gmock/gmock.h"
#include "gtest/gtest.h"
#include "kythe/cxx/common/path_utils.h"
#include "kythe/cxx/extractor/testlib.h"
#include "kythe/proto/analysis.pb.h"
#include "protobuf-matchers/protocol-buffer-matchers.h"

namespace kythe {
namespace {
using ::kythe::proto::CompilationUnit;
using ::protobuf_matchers::EquivToProto;
using ::testing::SizeIs;

constexpr absl::string_view kFilePath =
    "kythe/cxx/extractor/testdata/altroot/altpath/file.cc";

CompilationUnit ExpectedCompilation() {
  return ParseTextCompilationUnitOrDie(R"pb(
    v_name { language: "c++" }
    required_input {
      v_name { path: "altpath/file.cc" }
      info {
        path: "file.cc"
        digest: "a24091884bc15b53e380fe5b874d1bb52d89269fdf2592808ac70ba189204730"
      }
      details {
        [type.googleapis.com/kythe.proto.ContextDependentVersion] {
          row { source_context: "hash0" always_process: true }
        }
      }
    }
    argument: "/dummy/bin/g++"
    argument: "-target"
    argument: "dummy-target"
    argument: "-DKYTHE_IS_RUNNING=1"
    argument: "-resource-dir"
    argument: "/kythe_builtins"
    argument: "--driver-mode=g++"
    argument: "file.cc"
    argument: "-fsyntax-only"
    source_file: "file.cc"
    working_directory: "/root"
    entry_context: "hash0")pb");
}

// Verifies that the extractor properly handles KYTHE_ROOT_DIRECTORY
// other than the working directory.
TEST(RootDirectoryTest, AlternateRootDirectoryExtracts) {
  std::optional<std::string> resolved_path = ResolveRunfiles(kFilePath);
  ASSERT_TRUE(resolved_path.has_value());

  std::string filename(Basename(*resolved_path));
  std::string working_directory(Dirname(*resolved_path));
  std::string root_directory(Dirname(working_directory));

  ExtractorOptions extraction;
  extraction.working_directory = working_directory;
  extraction.arguments = {"--with_executable", "/dummy/bin/g++", filename};
  extraction.environment = {
      {"KYTHE_ROOT_DIRECTORY", root_directory},
  };

  CompilationUnit unit = ExtractSingleCompilationOrDie(std::move(extraction));
  ASSERT_THAT(unit.argument(), SizeIs(9));

  // Fix up things which may legitmately vary between runs.
  // TODO(shahms): use gMock protobuf matchers when available.
  CanonicalizeHashes(&unit);
  unit.set_argument(2, "dummy-target");
  unit.clear_details();

  EXPECT_THAT(unit, EquivToProto(ExpectedCompilation()));
}

// Verifies that the extractor properly picks a stable
// compilation unit working_directory that doesn't conflict
// with another compilation unit path.
TEST(RootDirectoryTest, WorkingDirectoryAvoidsConflict) {
  std::optional<std::string> resolved_path = ResolveRunfiles(kFilePath);
  ASSERT_TRUE(resolved_path.has_value());

  std::string filename(Basename(*resolved_path));
  std::string working_directory(Dirname(*resolved_path));
  std::string root_directory(Dirname(working_directory));

  ExtractorOptions extraction;
  extraction.working_directory = working_directory;
  extraction.arguments = {
      "--with_executable", "/dummy/bin/g++",
      // The extractor should avoid using a stable working_directory if the
      // actual working_directory is present in the argument list.
      absl::StrCat("-DUNUSED=", working_directory), filename};
  extraction.environment = {
      {"KYTHE_ROOT_DIRECTORY", root_directory},
  };

  CompilationUnit unit = ExtractSingleCompilationOrDie(std::move(extraction));
  ASSERT_THAT(unit.argument(), SizeIs(10));

  // Fix up things which may legitmately vary between runs.
  // TODO(shahms): use gMock protobuf matchers when available.
  CanonicalizeHashes(&unit);
  unit.set_argument(2, "dummy-target");
  unit.clear_details();

  auto expected = ExpectedCompilation();
  expected.add_argument(absl::StrCat("-DUNUSED=", working_directory));
  // RepeatedPtrField lacks an "insert" method, so we have to use rotate
  // to position it correctly.
  std::rotate(expected.mutable_argument()->pointer_begin() + 7,
              expected.mutable_argument()->pointer_end() - 1,
              expected.mutable_argument()->pointer_end());
  expected.set_working_directory(working_directory);

  EXPECT_THAT(unit, EquivToProto(expected));
}

}  // namespace
}  // namespace kythe
